lean_source/06_Abstract_Algebra/source_01_Structures.lean

```diff
@@ -0,0 +1,503 @@
+(preserves_mul : ∀ x c, f (c * x) = c * f x)
+
+section
+variables (f : ℝ → ℝ) (linf : is_linear f)
+
+#check linf.is_additive
+#check linf.preserves_mul
+
+end
+-- QUOTE.
+
+/- TEXT:
+It is worth pointing out that structures are not the only way to bundle
+together data.
+The ``point`` data structure can be defined using the generic type product,
+and ``is_linear`` can be defined with a simple ``and``.
+EXAMPLES: -/
+-- QUOTE:
+def point'' := ℝ × ℝ × ℝ
+
+def is_linear' (f : ℝ → ℝ) :=
+(∀ x y, f (x + y) = f x + f y) ∧ (∀ x c, f (c * x) = c * f x)
+-- QUOTE.
+
+/- TEXT:
+Generic type constructions can even be used in place of structures
+with dependencies between their components.
+For example, the *subtype* construction combines a piece of data with
+a property.
+You can think of of the type ``preal`` in the next example as being
+the type of positive real numbers.
+Any ``x : preal`` has two components: the value, and the property of being
+positive.
+You can access these components as ``x.val``, which has type ``ℝ``,
+and ``x.property``, which represents the fact ``0 < x.val``.
+EXAMPLES: -/
+-- QUOTE:
+def preal := { y : ℝ // 0 < y }
+
+section
+variable x : preal
+
+#check x.val
+#check x.property
+
+#check x.1
+#check x.2
+
+end
+-- QUOTE.
+
+/- TEXT:
+We could have used subtypes to define the standard 2-simplex,
+as well as the standard :math:`n`-simplex for an arbitrary :math:`n`.
+EXAMPLES: -/
+-- QUOTE:
+def standard_two_simplex' :=
+{ p : ℝ × ℝ × ℝ // 0 ≤ p.1 ∧ 0 ≤ p.2.1 ∧ 0 ≤ p.2.2 ∧ p.1 + p.2.1 + p.2.2 = 1 }
+
+def standard_simplex' (n : ℕ) :=
+{ v : fin n → ℝ // (∀ i : fin n, 0 ≤ v i) ∧ (∑ i, v i = 1) }
+-- QUOTE.
+
+/- TEXT:
+Similarly, *Sigma types* are generalizations of ordered pairs,
+whereby the type of the second component depends on the type of
+the first.
+EXAMPLES: -/
+-- QUOTE:
+def std_simplex := Σ n : ℕ, standard_simplex n
+
+section
+variable s : std_simplex
+
+#check s.fst
+#check s.snd
+
+#check s.1
+#check s.2
+
+end
+-- QUOTE.
+
+/- TEXT:
+Given ``s : std_simplex``, the first component ``s.fst`` is a natural
+number, and the second component is an element of the corresponding
+simplex ``standard_simplex s.fst``.
+The difference between a Sigma type and a subtype is that
+the second component of a Sigma type is data rather than a proposition.
+
+But even though we can use products, subtypes, and Sigma types
+instead of structures, using structures has a number of advantages.
+Defining a structure abstracts away the underlying representation
+and provides custom names for the functions that access the components.
+This makes proofs more robust:
+proofs that rely only on the interface to a structure
+will generally continue to work when we change the definition,
+as long as we redefine the old accessors in terms of the new definition.
+Moreover, as we are about to see, Lean provides support for
+weaving structures together into a rich, interconnected hierarchy,
+and for managing the interactions between them.
```

Comment:
We could make this paragraph much less abstract by showing how to access the components of a point with the definition `def point'' := ℝ × ℝ × ℝ`. However if we do that it would probably be honest to also mention the possibility of using `fin 3 → ℝ` as the definition. This interesting anyhow, because I think very few mathematician realize that defining ℝ^n as an iterated cartesian product is a polite lie that would be a nightmare if taken seriously.

Reply:
Also left as a TODO. Let's see how it fits with the next section.